content/blog/2024-10-21-testing-the-limits/index.md

```diff
@@ -0,0 +1,239 @@
+
+![Test architecture showing two nodes: a Benchmark server with the ip of 10.0.11.221 with an arrow pointing to a Valkey server with an ip of 10.0.1.136](images/test_setup.png)
+
+## Setting up our environment
+
+Below is a straightforward docker compose file that will start a single Valkey container. This container will bind Valkey to port 6379 on the host device. This means that is exposed to anyone with access to your network! This is important for us to access it from the benchmarking server. 
+
+```yaml
+# valkey.yaml
+services:
+  valkey-1:
+    image: valkey/valkey:latest
+    hostname: valkey1
+    command: valkey-server --port 6379 --requirepass ${VALKEY_PASSWORD}
+    volumes:
+      - ./data:/data
+    network_mode: host
+
+volumes:
+  data:
+    driver: local
+```
+
+Since we are exposing this to our internal network we will create a password for our default user. I used `head -16 /dev/urandom | openssl sha1` to generate a random password. Because of how fast Valkey can process requests a brute force attack could try hundereds of thousands of passwords per second. After generating that password I put it in a `.env` file in the same directory as our docker compose. 
+
+```bash
+#.env
+VALKEY_PASSWORD=e41fb9818502071d592b36b99f63003019861dad
+```
+
+Now by running `docker compose -f valkey.yaml up -d` Valkey server will start with password we set! 
+
+## Baseline test
+
+Now we are ready to do some baseline testing. We will log into the benchmarking server. If you haven't installed redis-benchmark yet you can do so with `sudo apt install redis-tools`.  
+
+```bash
+redis-benchmark -n 1000000 -t set,get -P 16 -q -a <PASSWORD FROM .env>  -h 10.0.1.136
+```
+
+Test breakdown:
+- `-n` - this will run 1,000,000 operations using the commands set in -t
+- `-t` - will run the set and get tests
+- `-P` - this specifies that we would like the tests to use 16 pipelines (send 16 operations per request). 
+- `-q` - silences the output to show only the final results
+- `-a` - use the specified password
+- `-h` - run the test against the specified host
+
+Honestly, I was asonished by the first set of results I got. Sure, I expected Valkey to be fast but this speed from a single board computer?? ON A SINGLE THREAD?? Its amazing. 
+
+```bash
+redis-benchmark -n 1000000 -t set,get -P 16 -q -a e41fb9818502071d592b36b99f63003019861dad  -h 10.0.1.136
+SET: 173040.33 requests per second, p50=4.503 msec                    
+GET: 307031.00 requests per second, p50=2.455 msec
+```
+
+Between the two tests we averaged 240,000 requests per second with a median time of 3.45 msec (or 0.0034 seconds) for each operation. 
```

Comment:
```suggestion
Between the two tests we averaged 240,000 requests per second with a median time of 3.45 usec (or 0.0000034 seconds) for each operation. 
```
Order of magnitude off :)

Reply:
Ahhh I see what you mean. Ill clarify the for each operation to say per round trip for the pipeline batch. 

As for your other comment you mean it might not be worth including in the blog at all? I am happy to remove it as it scales with the RPS.

Reply:
> As for your other comment you mean it might not be worth including latency in the blog at all? I am happy to remove it as it scales with the RPS.

It actually doesn't scale, it's just an artifact of the way the benchmark works. The amount of time a single command takes is a constant (more or less it's just a hashmap lookup) +  how much queuing is present on the server. Queuing artificially decreases as the server is able to sustain more load if you have the same number of pending clients since the benchmark sends requests as fast as possible. This isn't how real world scenarios work. Your application sends a finite number of requests per second. If you sent a finite 100k RPS, the latency would be identical on both setups, since neither system would queue. What you actually want to generate is a curve which is the "max throughput" for a given latency, but that requires a highly adaptive system.

tl;dr, the benchmark is built to really only test throughput. If you want to effectively test latency you need to bound the throughput.